lib/config.go

```diff
@@ -5,81 +5,47 @@ import (
 	"fmt"
 
 	"github.com/ghodss/yaml"
-	golog "github.com/ipfs/go-log"
-	"github.com/qri-io/ioes"
 	"github.com/qri-io/qri/config"
-	"github.com/qri-io/qri/config/migrate"
 )
 
-var (
-	// Config is the global configuration object
-	Config *config.Config
-	// ConfigFilepath is the default location for a config file
-	ConfigFilepath string
-)
-
-// SaveConfig is a function that updates the configuration file
-var SaveConfig = func() error {
-	if err := Config.WriteToFile(ConfigFilepath); err != nil {
-		return fmt.Errorf("error saving profile: %s", err)
-	}
-	return nil
+// NewConfigMethods creates a configuration handle from an instance
+func NewConfigMethods(inst Instance) ConfigMethods {
+	return ConfigMethods{Instance: inst}
 }
 
-// LoadConfig loads the global default configuration
-func LoadConfig(streams ioes.IOStreams, path string) (err error) {
-	cfg, err := config.ReadFromFile(path)
-	if err != nil {
-		return err
-	}
-
-	if cfg.Profile == nil {
-		err = fmt.Errorf("missing profile")
-		return err
-	}
-
-	// configure logging straight away
-	if cfg.Logging != nil {
-		for name, level := range cfg.Logging.Levels {
-			golog.SetLogLevel(name, level)
-		}
-	}
-
-	Config = cfg
-
-	migrated, err := migrate.RunMigrations(streams, cfg)
-	if err != nil {
-		return err
-	}
-
-	if migrated {
-		return SaveConfig()
-	}
-
-	return nil
+// ConfigMethods encapsulates changes to a qri configuration
+type ConfigMethods struct {
+	Instance
```

Comment:
In my humble opinion, the biggest win we want to accomplish with this change is consolidating all of the state that qri uses into a single struct. Moving the Config from a global object into part of this struct is part of accomplishing this goal. All of our private code (inside and under lib/) should be using this struct and passing it around to other parts of the code. However, I don't think it needs to be an interface that hides the details, since it's only for internal use. With that in mind, I think the right approach looks like this:

```
type instance struct {
  cfg *Config
  node *p2p.QriNode
  ctx Context
}

type ConfigMethods struct {
  instance
}

func (c *ConfigMethods) SomeFunc() {
  c.node.Connect() // or whatever
}
```

When we make a public facing interface for uses of lib/, they can get something that hides these details and doesn't give direct access to these internal fields.

Reply:
@dustmop & I have been discussing this at length offline. I'm very much won over by this approach, especially when combined with some of the other feedback in this review (like a convenience function for building test `instance`s).

I'm wondering if we should take it one step further: make `instance` an exported struct with no exported fields:

```go
type Instance struct {
  cfg *Config
  node *p2p.QriNode
  ctx Context
}
```

This PR currently has `lib` exporting an `Instance` interface:
```go
// Instance is the interface that bundles the foundational values of a
// qri instance. Instance provides the basis for creating Method constructors,
// which actually do qri things
// think of instance as the "core" of the qri ecosystem
type Instance interface {
	// Context returns the base context this instance is using. Any resources
	// built from this instance should inherit from this context and obey
	// calls from the ctx.Done(), releasing any & all resources
	Context() context.Context
	// Teardown closes the instance by closing the base context
	Teardown()
	// Config returns the current configuration for this
	Config() *config.Config
	Node() *p2p.QriNode
	Repo() repo.Repo
	RPC() *rpc.Client
}
```
The initial thinking is defaulting to being defensive: by passing an interface we're making fewer promises to userland. I've spent some time asking the question: _when would we want multiple implementations of this interface?_ And can't find any reasonable situations where this would be the case.

Combining that with @dustmop's feedback, if `interface` were exported we'd have two nice properties. First, code like this now looks far more ok in `lib` tests because Instance is exported, and lib can access the unexported fields:
```go
inst := &Instance{ node: node }
```

Secondly, we can take Dustmop's suggestion one step further now that Instance is a concrete type, having "Method Types" build directly on an instance if they require no additional fields:
```go
type ConfigMethods Instance

func (c *ConfigMethods) SomeFunc() {
  c.node.Connect() // or whatever
}
```
This is only possible if Instance isn't an interface type.